### PR TITLE
FIX: data.shape[2] may be wrongly estimated in _BaseEpochs.__init__()

### PR DESCRIPTION
**Pull request by cmoutard**
In \__init\__() within _BaseEpochs, data.shape[2] is evaluated as follow :
```
round((tmax - tmin) * self.info['sfreq']) + 1
```

And if data.shape[2] is not egal to this estimation, it throws an error :
```
raise RuntimeError('bad data shape')
```

However, in some cases, I think the estimation is wrong. Here is an exemple :
Let's say that originally tmin=0.0 and tmax=0.009 (epochs lasting 9ms), and that self.info['sfreq'] was fixed to 500.
Then, after decimation, times and data.shape[2] would be : 

```
epochs.times = [0., 0.002, 0.004, 0.006, 0.008]        
data.shape[2] = 5
```

But :
```
(tmax - tmin) * self.info['sfreq'] = 0.009*500 = 4.5
round(4.5) = 5
```
So :
```
round((tmax - tmin) * self.info['sfreq']) + 1 = 6
```

\__init\__() in _BaseEpochs whould then throw an error while everything is fine.

To sum it up, when the first decimal of (tmax - tmin) * self.info['sfreq'] is >=5, the rounding and the +1 are redondant, making the estimate of data.shape[2] wrong.

I propose here to fix it by replacing _round_ by _int_, to floor the float and suppress the redondant increment. 

Btw, this issue is fairely related to another problem I met with tmin and tmax definition + decimation, which is still out there (see PR #2486, closed without being merged...).

### CODE REVIEW

**Comment by cmoutard** [comment]
New fail with Travis with xdawn.
Waiting for your feedback before going on...

**Comment by kingjr** [comment]
This seems also linked to #2645

**Comment by jaeilepp** [comment]
I have a hard time reproducing this problem. ``round(4.5)`` actually returns 4 and I suppose it is because 
> it’s a result of the fact that most decimal fractions can’t be represented exactly as a float.

Then again if you modify the first example to go up to 0.011 you get:
``epochs.times = [ 0.   ,  0.002,  0.004,  0.006,  0.008,  0.01 ,  0.012]``
and ``round(5.5) = 6``.
So without digging any deeper, it seems that rounding works the same way both directions.

File: mne/epochs.py [hunk 1]
```diff
@@ -278,7 +278,7 @@ def __init__(self, info, data, events, event_id, tmin, tmax,
         else:
             assert decim == 1
             if data.ndim != 3 or data.shape[2] != \
-                    round((tmax - tmin) * self.info['sfreq']) + 1:
+                    int((tmax - tmin) * self.info['sfreq']) + 1:
```
Comment (dengemann) [inline mne/epochs.py]: maybe it could be replaced and use our safe mne.utils._time_mask function

File: mne/tests/test_epochs.py [hunk 2]
```diff
@@ -905,7 +905,7 @@ def test_crop():
     assert_array_equal(data3, data_normal[:, :, tmask])
 
     # test time info is correct
-    epochs = EpochsArray(np.zeros((1, 1, 1000)), create_info(1, 1000., 'eeg'),
+    epochs = EpochsArray(np.zeros((1, 1, 1001)), create_info(1, 1000., 'eeg'),
```
Comment (agramfort) [inline mne/tests/test_epochs.py]: the test fails if you don't update this line?

or it's to test the fix?

if it's the latter then you should have both tests

Comment (cmoutard) [inline mne/tests/test_epochs.py]: the test does fail if I don't update this line

Comment (agramfort) [inline mne/tests/test_epochs.py]: then it's not the right fix. What if I have 1000 samples and I want to call epochs.crop(-.200, .700) ?

it should work all the time.

Comment (cmoutard) [inline mne/tests/test_epochs.py]: Ok, I think I get the problem now

**Comment by cmoutard** [comment]
Yes, you're right

**Comment by Eric89GXL** [comment]
Yeah pretty sure the round() is only there because floating point arithmetic is not perfect. In other words, we should always be very close to an integer after the division, and the `round` should just touch it a tiny bit in the right direction. If that's not the case, there is a bug somewhere else with how we're choosing tmin/tmax etc.

**Comment by cmoutard** [comment]
Ok, so we loop back to  #2486 => when one decimate, the original tmax is kept while the effective one is possibily changed (epochs.times[-1]). So this may conduced to the situation I am in...
I would plead again to allow the update of tmax.
I have been told that nobody should decimate final results.
But I have large epochs and numerous decoders => I have no other choice than decimation for volume disk and time issues, even for the final figures... 
Is there any way we can consider to make tmax more flexible ? This would save me a lot of ugly local fixes...

**Comment by Eric89GXL** [comment]
> Ok, so we loop back to #2486 => when one decimate, the original tmax is kept while the effective one is possibily changed (epochs.times[-1]).

I think by this you mean the zero-point is kept fixed, while the start (tmin) and end (tmax) might change due to decimation?

> I have been told that nobody should decimate final results.

AFAIK there's nothing wrong with decimating epochs if your raw data that were used to create Epochs were sufficiently low-passed. Or did you mean something else?

> I would plead again to allow the update of tmax after decimation... Is there any way we can consider to make tmax more flexible ? This would save me a lot of ugly local fixes...

The question in my mind is (as it was in #2486), what are you trying to accomplish by changing `tmax`? If your data are sufficiently low-passed to justify the decimation, then you won't really have the resolution to get really meaningful differences between what `tmax` gets set to, and whatever you'd want it to be through a time-shift procedure like #2486. Is it more than an aesthetic plotting thing? If so, you can upsample your data properly with a `sinc` for plotting purposes to make things appear properly smoothed (as opposed to jagged like you might get from standard linear interpolation between points).

**Comment by agramfort** [comment]
travis is not happy

can you also test for both 1000 and 1001 ?

**Comment by jaeilepp** [comment]
I think the problem is that if you resample, the rounding isn't guaranteed to work to the right direction anymore. I think we could set the resampling to return uneven number of samples if the rounding doesn't work in the right direction. This might seem a bit hackish, but I think it's the best solution without changing the ``tmax``.

The only "unexpected" result would be that the number of samples wouldn't always be the expected amount. For instance, in https://github.com/mne-tools/mne-python/issues/2645 the epochs instances would have 11 samples. Sorry, but I don't know how to explain it any better. I can make a PR if you think this is the way to go.

**Comment by agramfort** [comment]
not sure to follow but this needs a fix...

**Comment by Eric89GXL** [comment]
> I think the problem is that if you resample, the rounding isn't guaranteed to work to the right direction anymore. I think we could set the resampling to return uneven number of samples if the rounding doesn't work in the right direction. This might seem a bit hackish, but I think it's the best solution without changing the tmax.

I think the correct solution is to change the `tmax`. When you decimate, either your `tmax` or `tmin` (or both) will have to change. It makes sense to me to change `tmax`, since typically decimation works like `data[::decim_amount]`, i.e. taking the zeroth and every nth samplethereafter.

**Comment by Eric89GXL** [comment]
I can come up with a fix in a bit unless someone else wants to.

**Comment by agramfort** [comment]
go for it

**Comment by Eric89GXL** [comment]
Closing for #2645
